server.go

```diff
@@ -1039,7 +1051,15 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 	}
 	opts := &transport.Options{Last: true}
 
-	if err := s.sendResponse(t, stream, reply, cp, opts, comp); err != nil {
+	f, err := s.sendResponse(t, stream, reply, cp, opts, comp, sh == nil && binlog == nil)
+	if f != nil {
+		defer func() {
+			// Do not wait for the unary response to be written to the wire. Spawn a
+			// goroutine to wait on the waitgroup instead to be non-blocking.
+			go waitCallReturnBuffer(f, opts.ReturnBufferWaitGroup)
```

Comment:
When `Done()` is called by loopy, it doesn't mean that we can free the buffer. On the client side, if a future message fails, every message needs to be retried -- even ones that were successfully written to the wire. So we'll need to keep an unreturned reference to these payloads so that they aren't recycled (which will corrupt the data).

For this, we need to wait till the RPC is committed so that it won't be retried. But that alone doesn't guarantee that these byte slices can be returned -- what if there are messages in loopy's `cbuf` that haven't been written to the wire? IIUC, a query is defined to be committed if it receives headers back from the server; in a bidirectional stream, a client may receive headers from the server before all client streaming messages are written out. We need to wait till both of these conditions are satisfied:

1. The RPC is committed (i.e. no more retries) -- which means we can't call `f()` immediately within `Done()` even if the counter is 0 because it may be retried in the future.
2. All outgoing messages have been written to the wire (i.e. no more references to the payload).

The issue now is that this may be satisfied after the end of `processUnaryRPC` if loopy has a large enough backlog to process.

Reply:
And AFAICT the only way to be able to do stuff well after the lifetime of an RPC call is by forking a goroutine.

Reply:
Can we make it so that `cs.commitAttempt` and loopy finishing writing each do a `Done`?  `Add(2)` first, and when the both conditions are satisfied, we will return the buffer?

Reply:
Nice, I think that would work. It'd need to be a `Add(1)` for each attempt + `Add(1)` for loopy enqueue, but I see what you mean. Thanks, I'll try this!

Reply:
When we commit the attempt, we never retry, so if we Done() in `commitAttempt` I think we should be good.  (I'm not sure if we call `commitAttempt` unconditionally; if not, we'd want to `Add(1)` instead of `2` if this RPC will not be retried.)

Reply:
Done. See PATCH V3.

Reply:
one sec, need to fix something, WIP for a few minutes, sorry

Reply:
@dfawley ok, this is once again ready for review.